packages/eslint-plugin/src/rules/no-misused-object-likes.ts

```diff
@@ -0,0 +1,149 @@
+import type { Identifier, MemberExpression } from '@typescript-eslint/ast-spec';
+import type { TSESTree } from '@typescript-eslint/utils';
+
+import { createRule, getParserServices } from '../util';
+
+export type Options = [
+  {
+    checkObjectKeysForMap?: boolean;
+    checkObjectValuesForMap?: boolean;
+    checkObjectEntriesForMap?: boolean;
+    checkObjectKeysForSet?: boolean;
+    checkObjectValuesForSet?: boolean;
+    checkObjectEntriesForSet?: boolean;
+  },
+];
+export type MessageIds =
+  | 'objectKeysForMap'
+  | 'objectValuesForMap'
+  | 'objectEntriesForMap'
+  | 'objectKeysForSet'
+  | 'objectValuesForSet'
+  | 'objectEntriesForSet';
+
+export default createRule<Options, MessageIds>({
+  name: 'no-misused-object-likes',
+  defaultOptions: [
+    {
+      checkObjectKeysForMap: true,
+      checkObjectValuesForMap: true,
+      checkObjectEntriesForMap: true,
+      checkObjectKeysForSet: true,
+      checkObjectValuesForSet: true,
+      checkObjectEntriesForSet: true,
+    },
+  ],
+
+  meta: {
+    type: 'problem',
+    docs: {
+      description:
+        'Enforce check `Object.values(...)`, `Object.keys(...)`, `Object.entries(...)` usage with Map/Set objects',
+      requiresTypeChecking: false,
+    },
+    messages: {
+      objectKeysForMap: "Don't use `Object.keys()` for Map objects",
```

Comment:
I think this may be overly specific, granularity for `SetLike` and `MapLike` is likely sufficient since it would be weird to allow calling `Object.values(` on a Map but banning `Object.keys`